### PR TITLE
GPFlow 2.0: ARD and Stationaries Refactor

### PR DESCRIPTION
**Pull request by jahall**
Migrating the generalized periodic enhancement from 1.0. This is the initial groundwork to make that happen i.e. addresses the issues outlined in https://github.com/GPflow/GPflow/issues/1077. In addition some minor refactoring has been done e.g. to move Periodic out of misc.py and into periodic.py.

Will re-introduce the actual enhancement in a follow up commit tomorrow.

### CODE REVIEW

**Comment by jahall** [comment]
@st-- Would welcome your feedback on the initial groundwork.

**Comment by jahall** [comment]
@st-- May need your help tracking down the bugs thrown up in CI. I think they are related to my removal of the `ard` kwarg and associated ard validation check.

File: gpflow/kernels/base.py [hunk 1]
```diff
@@ -117,6 +120,19 @@ def slice_cov(self, cov: tf.Tensor) -> tf.Tensor:
...
+    def validate_ard_parameter(self, X, parameter):
+        """
+        Check that the input dimension matches the dimension of the parameter.
+        """
+        if tf.rank(parameter) == 0:
+            return
+        X_shape = tf.shape(X)
+        l_shape = tf.shape(parameter)
+        if l_shape[-1] != X_shape[-1]:
+            # TODO: Maybe this should be a tf.errors.InvalidArgumentError
```
Comment (st--) [inline gpflow/kernels/base.py]: I don't think we should use another package's exceptions, especially if they're that unspecific...

**Comment by st--** [comment]
@jahall thanks for your work on this. I think (at least some of) the tests fail because the `ard` flag is actually used in the kernel expectations (`expectations/`). We could keep an `ard` property that checks whether the lengthscale parameter is a scalar (ard=False) or a vector (ard=True)...

I'm wondering if it'd be cleaner (& easier to review) to move the refactoring of ARD and `K_r2`/`K_r` into its own PR?

File: gpflow/kernels/base.py (hunk 1 above)
Comment (jahall) [reply gpflow/kernels/base.py]: Agreed - will remove comment

File: gpflow/kernels/linears.py [hunk 2]
```diff
@@ -13,24 +13,30 @@ class Linear(Kernel):
     where σ²  is the variance parameter.
     """
 
-    def __init__(self, variance=1.0, active_dims=None, ard=None):
+    def __init__(self, variance=1.0, active_dims=None):
         """
         - input_dim is the dimension of the input to the kernel
         - variance is the (initial) value for the variance parameter(s)
-          if ard=True, there is one variance per input
+          defaults to 1.0 but supports ARD i.e. one variance per input.
```
Comment (st--) [inline gpflow/kernels/linears.py]: Could we have a more complete explanation? e.g. if you want ARD but initialised to 1 you'd have to pass in [1, 1, ..., 1] (where the length of the array is equal to the input dimension)

File: gpflow/kernels/base.py [hunk 3]
```diff
@@ -117,6 +120,24 @@ def slice_cov(self, cov: tf.Tensor) -> tf.Tensor:
 
         return cov
 
+    def validate_ard_parameter(self, X, parameter):
```
Comment (st--) [inline gpflow/kernels/base.py]: Could also check whether `active_dims` has the right length (if not None)?

File: gpflow/kernels/linears.py (hunk 2 above)
Comment (st--) [reply gpflow/kernels/linears.py]: (Applies to all kernels with ard feature)

File: gpflow/kernels/stationaries.py [hunk 4]
```diff
@@ -17,48 +17,54 @@ class Stationary(Kernel):
...
     """
 
-    def __init__(self, variance=1.0, lengthscale=1.0, active_dims=None, ard=None):
+    def __init__(self, variance=1.0, lengthscale=1.0, active_dims=None):
         """
         - input_dim is the dimension of the input to the kernel
         - variance is the (initial) value for the variance parameter
         - lengthscale is the initial value for the lengthscale parameter
-          defaults to 1.0 (ard=False) or np.ones(input_dim) (ard=True).
+          defaults to 1.0.
```
Comment (st--) [inline gpflow/kernels/stationaries.py]: explain how to get ard behaviour?

File: gpflow/kernels/stationaries.py [hunk 5]
```diff
@@ -17,48 +17,54 @@ class Stationary(Kernel):
...
 
     def K_diag(self, X, presliced=False):
         return tf.fill((X.shape[:-1]), tf.squeeze(self.variance))
 
-    def K_r(self, r):
+    def K_r2(self, r2):
         """
-        Returns the kernel evaluated on `r`, which is the scaled Euclidean distance
-        Should operate element-wise on r
+        Returns the kernel evaluated on `r²`, which is the squared scaled Euclidean distance
```
Comment (st--) [inline gpflow/kernels/stationaries.py]:
```suggestion
        Returns the kernel evaluated on r² (`r2`), which is the squared scaled Euclidean distance
```

File: tests/test_features.py [hunk 6]
```diff
@@ -30,9 +30,8 @@ def test_inducing_points_inducing_variable_len(N, D):
 
 _kernel_setups = [
     gpflow.kernels.SquaredExponential(variance=0.46,
-                                      lengthscale=np.random.uniform(0.5, 3., 5),
-                                      ard=True),
-    gpflow.kernels.Periodic(period=0.4, variance=1.8)
+                                      lengthscale=np.random.uniform(0.5, 3., 5)),
```
Comment (st--) [inline tests/test_features.py]: should the lengthscale be an array to ensure it's actually testing ard=True ?

File: tests/test_kernels.py [hunk 7]
```diff
@@ -137,15 +132,14 @@ def test_arccosine_non_implemented_order(order):
         gpflow.kernels.ArcCosine(order=order)
 
 
-@pytest.mark.parametrize('ard', [True, False])
```
Comment (st--) [inline tests/test_kernels.py]: I would keep this in - maybe as
```python
@pytest.mark.parametrize('weight_variances', [
    1.,  # ard = False
    [1., 1., 1.],  # ard = True
])
```
(and then take `weight_variances` out of the following parametrization)
(assumes input_dim=3)

File: tests/test_kernels.py [hunk 8]
```diff
@@ -165,8 +159,6 @@ def _assert_periodic_kern_err(lengthscale, variance, period, X):
     gram_matrix = kernel(X)
     reference_gram_matrix = _ref_periodic(X, lengthscale, variance, period)
 
-    assert_allclose(gram_matrix, reference_gram_matrix)
```
Comment (st--) [inline tests/test_kernels.py]: why did this get removed?

File: tests/test_kernels.py (hunk 7 above)
Comment (st--) [reply tests/test_kernels.py]: or did this already get tested elsewhere ? :thinking:

File: tests/test_kernels.py [hunk 9]
```diff
@@ -243,7 +235,7 @@ def test_coregion_slice(N, input_dim, output_dim, rank):
 kernel_setups_extended = kernel_setups + [
     SquaredExponential() + Linear(),
     SquaredExponential() * Linear(),
-    SquaredExponential() + Linear(ard=True, variance=rng.rand(_dim, 1).reshape(-1))
+    SquaredExponential() + Linear(variance=rng.rand(_dim, 1).reshape(-1))
```
Comment (st--) [inline tests/test_kernels.py]: Why not
```suggestion
    SquaredExponential() + Linear(variance=rng.rand(_dim))
```
?

File: tests/test_kernels.py [hunk 10]
```diff
@@ -362,7 +354,7 @@ def test_active_product(N, D):
         np.hstack([ls[:rand_idx], ls[rand_idx + 1:]]), ls[rand_idx], ls
     ]
     kernels = [
-        gpflow.kernels.SquaredExponential(lengthscale=lengthscale, active_dims=dims, ard=True)
```
Comment (st--) [inline tests/test_kernels.py]: does ard=True still get tested?

File: tests/test_kernels.py [hunk 11]
```diff
@@ -389,23 +381,9 @@ def test_ard_init_scalar(D):
...
         k1(rng.randn(N, 4))
-    with pytest.raises(tf.errors.InvalidArgumentError):
-        k2 = gpflow.kernels.SquaredExponential(lengthscale=np.ones(3), ard=ard)
+    with pytest.raises(ValueError):
+        k2 = gpflow.kernels.SquaredExponential(lengthscale=np.ones(3))
         k2(rng.randn(N, 2))
-
-
-@pytest.mark.parametrize('D', [4, 7])
-def test_ard_init_MLP(D):
```
Comment (st--) [inline tests/test_kernels.py]: why did this disappear?

File: gpflow/kernels/linears.py (hunk 2 above)
Comment (jahall) [reply gpflow/kernels/linears.py]: Makes sense - have done this

File: gpflow/kernels/base.py (hunk 3 above)
Comment (jahall) [reply gpflow/kernels/base.py]: I will add in a check on __init__ to ensure size of `active_dims` has the right length (if not None and if not a slice)

File: tests/test_kernels.py (hunk 9 above)
Comment (jahall) [reply tests/test_kernels.py]: Ha - yeah I'm not sure why that was defined that way. I'll change it.

File: tests/test_kernels.py (hunk 11 above)
Comment (jahall) [reply tests/test_kernels.py]: This disappeared because the test is no longer valid - it was checking that if you passed in a scalar `weight_variances` with `ard=True` that it got expanded to an array. We now rely on the user explicitly specifying an array.

File: tests/test_kernels.py (hunk 7 above)
Comment (jahall) [reply tests/test_kernels.py]: Actually - this is now sufficiently tested in `test_arccosine_kern_err` - this test can be removed

**Comment by jahall** [comment]
@st-- Have addressed all comments. However, I cannot for the life of me figure out how to get this `validate_ard_parameter` method to work properly as `tf.cond` doesn't always lazily evaluate :( - see TODO in code. Other than that I think this is in reasonable shape. Just fixing the notebooks.

File: gpflow/kernels/base.py [hunk 12]
```diff
@@ -117,6 +121,42 @@ def slice_cov(self, cov: tf.Tensor) -> tf.Tensor:
...
+        is_ard = isinstance(raw_parameter, Iterable)
+        if (is_ard
+                and self.active_dims is not None
+                and not isinstance(self.active_dims, slice)
+                and len(raw_parameter) != len(self.active_dims)):
+            raise ValueError(f"Dimension of `active_dims` {self.active_dims} does not match "
+                             f"dimension of the ard parameter ({len(raw_parameter)})")
+        return is_ard
+
+    def validate_ard_parameter(self, X, parameter):
```
Comment (awav) [inline gpflow/kernels/base.py]: This is wierd function, it just returns True? Why did you add this TODO and commented out code?

File: tests/test_kernels.py [hunk 13]
```diff
@@ -380,32 +364,38 @@ def test_ard_init_scalar(D):
...
-    variances_1 = kernel_1.weight_variances.read_value()
-    variances_2 = kernel_2.weight_variances.read_value()
-    assert np.allclose(variances_1, variances_2, atol=1e-10)
+def test_ard_invalid_active_dims():
+    msg = r"Dimension of `active_dims` \[1\] does not match dimension of the ard parameter \(2\)"
+    with pytest.raises(ValueError, match=msg):
+        gpflow.kernels.SquaredExponential(lengthscale=np.ones(2), active_dims=[1])
+
+
+#@pytest.mark.parametrize('N', [4, 7])
```
Comment (awav) [inline tests/test_kernels.py]: Why this part is commented out?

Comment (jahall) [reply tests/test_kernels.py]: This is the associated test for the aforementioned commented code.

File: gpflow/kernels/linears.py [hunk 14]
```diff
@@ -13,24 +13,23 @@ class Linear(Kernel):
...
+            to induce ARD behaviour this must be initialised as an array the same
+            length as the the number of active dimensions e.g. [1., 1., 1.]
+        :param active_dims: a slice or list specifying which columns of X are used
         """
         super().__init__(active_dims)
-
-        # variance, self.ard = self._validate_ard_shape("variance", variance, ard)
-        self.ard = ard
         self.variance = Parameter(variance, transform=positive())
+        self.ard = self._is_ard(variance)
```
Comment (st--) [inline gpflow/kernels/linears.py]: I think it's better as a property (as you had before), e.g. if someone decides to replace the `variance` attribute with a different Parameter object. Same applies to the other kernels as well of course!

Comment (jahall) [reply gpflow/kernels/linears.py]: Good point - I have reintroduced `ard` as a parameter and improved the implementation by creating an `ARDKernel` base class to enable better code re-use.

File: gpflow/kernels/base.py (hunk 12 above)
Comment (jahall) [reply gpflow/kernels/base.py]: I have removed this altogether for now as I cannot figure out how to perform this validation correctly during graph execution.

**Comment by awav** [comment]
Hello @jahall , I apologize for closing your PR. It was closed automatically when I moved `develop-2.0` to the `develop`. Could you re-open the PR? Thanks!

**Comment by awav** [comment]
I think it is a bit of overkill to create a separate class because of the ard property of parameters (not even all of them) in a kernel :)

**Comment by jahall** [comment]
> I think it is a bit of overkill to create a separate class because of the ard property of parameters (not even all of them) in a kernel :)

That's reasonable. I'll refactor to make the change less intrusive.

File: gpflow/kernels/linears.py [hunk 15]
```diff
@@ -1,31 +1,36 @@
...
-        self.ard = ard
         self.variance = Parameter(variance, transform=positive())
+        self._validate_ard_active_dims(self.variance)
+
+    @property
+    def ard(self) -> bool:
+        """
+        Whether ARD behaviour is active.
+        """
+        return len(self.variance.shape) > 0
```
Comment (awav) [inline gpflow/kernels/linears.py]: self.variance.shape.ndims
```suggestion
        return self.variance.shape.ndims > 0
```

File: gpflow/kernels/misc.py [hunk 16]
```diff
@@ -55,10 +51,16 @@ def __init__(self,
...
         self.weight_variances = Parameter(weight_variances,
                                           transform=positive())
+        self._validate_ard_active_dims(self.weight_variances)
+
+    @property
+    def ard(self) -> bool:
+        """
+        Whether ARD behaviour is active.
+        """
+        return len(self.weight_variances.shape) > 0
```
Comment (awav) [inline gpflow/kernels/misc.py]:
```suggestion
        return self.weight_variances.shape.ndims > 0
```

**Comment by jahall** [comment]
@awav @st-- Any further changes required for this PR? I just ask as its been open over a month and contains necessary changes (i.e. the `K_r`/`K_r2` refactor in `stationaries`) to merge the generalized periodic kernel from v1 (see #1052)
